File: crabzilla-example1/crabzilla-example1-core/src/main/java/io/github/crabzilla/example1/customer/CustomerData.java
package io.github.crabzilla.example1.customer;

import io.github.crabzilla.core.DomainEvent;
import io.github.crabzilla.core.entity.EntityCommand;
import io.github.crabzilla.core.entity.EntityId;
import lombok.Value;

import java.time.Instant;
import java.util.UUID;

public class CustomerData {

  @Value
  public static class CustomerId implements EntityId {
    private String stringValue;

    @Override
    public String stringValue() {
      return stringValue;
    }
  }

  // tag::commands[]

  @Value
  public static class CreateCustomer implements EntityCommand {
    private UUID commandId;
    private CustomerId targetId;
    private String name;
  }

  @Value
  public static class ActivateCustomer implements EntityCommand {
    private UUID commandId;
    private CustomerId targetId;
    private String reason;
  }

  @Value
  public static class CreateActivateCustomer implements EntityCommand {
    private UUID commandId;
    private CustomerId targetId;
    private String name;
    private String reason;
  }

  // end::commands[]

  @Value
  public static class DeactivateCustomer implements EntityCommand {
    private UUID commandId;
    private CustomerId targetId;
    private String reason;
  }

  @Value
  public static class UnknownCommand implements EntityCommand {
    private UUID commandId;
    private CustomerData.CustomerId targetId;
  }

  // tag::events[]

  @Value
  public static class CustomerActivated implements DomainEvent {
    private String reason;
    private Instant when;
  }

  @Value
  public static class CustomerCreated implements DomainEvent {
    private CustomerId id;
    private String name;
  }

  @Value
  public static class CustomerDeactivated implements DomainEvent {
    private String reason;
    private Instant when;
  }

  // end::events[]

}

File: crabzilla-example1/crabzilla-example1-core/src/main/java/io/github/crabzilla/example1/customer/CustomerData.kt
package io.github.crabzilla.example1.customer

import io.github.crabzilla.core.Command
import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.EntityId
import java.time.Instant
import java.util.*

data class CustomerId(val id: String) : EntityId {
  override fun stringValue(): String {
    return id
  }
}

// tag::events[]

data class CustomerCreated(val id: CustomerId, val name: String) : DomainEvent

data class CustomerActivated(val reason: String, val _when: Instant) : DomainEvent

data class CustomerDeactivated(val reason: String, val _when: Instant) : DomainEvent

// end::events[]

// tag::commands[]

data class CreateCustomer(override val commandId: UUID, override val targetId: CustomerId,
                          val name: String) : Command

data class ActivateCustomer(override val commandId: UUID, override val targetId: CustomerId,
                            val reason: String) : Command

data class DeactivateCustomer(override val commandId: UUID, override val targetId: CustomerId,
                              val reason: String) : Command

data class CreateActivateCustomer(override val commandId: UUID,
                                  override val targetId: CustomerId,
                                  val name: String, val reason: String) : Command

// end::commands[]
